### PR TITLE
MOB-4382: Updated rubicon adapter to support sdk version, device pixel ratio, and user language. Initial commit to support mobile.

### PR DESCRIPTION
**Pull request by sandraleon**
…nd user language

### CODE REVIEW

**Comment by bretg** [comment]
@sandraleon - please provide a sample request with these new parameters.

File: adapters/rubicon/rubicon.go [hunk 1]
```diff
@@ -108,6 +114,18 @@ type rubiconTargetingObj struct {
 	Values []string `json:"values"`
 }
 
+type rubiconDeviceExtRP struct {
+	PixelRatio float64 `json:"pixelratio"`
```
Comment (DucChau) [inline adapters/rubicon/rubicon.go]: Is the json field name "pixelratio" consistent with how our full SDK handles it? or should it be labeled "pixel_ratio"?

Comment (sandraleon) [reply adapters/rubicon/rubicon.go]: "pixelratio" is correct. this is the key that the XAPI looks for.

**Comment by sandraleon** [comment]
@bretg here's a sample request:
{
	"user": {
		"age": 25,
		**"language": "arabic",**
		"gender": "F"
	},
	"max_key_length": 20,
	"account_id": "2763",
	"sort_bids": 1,
	"device": {
		"w": 320,
		"h": 568,
		"ifa": "FB37FC22-D6A1-41D8-B4A2-1AE2BE932CD7",
		"devtime": 1507845025,
		"osv": "11.0.2",
		"lmt": 1,
		"connectiontype": 1,
		"os": "iOS",
		"geo": {
			"lat": 50,
			"lon": 60
		},
		"make": "Apple",
		"ua": "Mozilla\/5.0 (iPhone; CPU iPhone OS 11_0_2 like Mac OS X) AppleWebKit\/604.1.38 (KHTML, like Gecko) Mobile\/15A421",
		"model": "iPhone6,1",
		"carrier": "T-Mobile",
		**"pxratio": 2**
	},
	"app": {
		"ver": "2.0.0",
		"bundle": "AppNexus.PrebidMobileDemo"
	},
	"cache_markup": 0,
	"ad_units": [{
		"sizes": [{
				"w": 320,
				"h": 50
			},
			{
				"w": 300,
				"h": 250
			}
		],
		"bids": [{
			"bidder": "rubicon",
			"params": {
				"zoneId": 743394,
				"siteId": 68780,
				"accountId": 2763
			}
		}],
		"code": "BannerAd"
	}],
	**"sdk": {
		"version": "0.1.1",
		"platform": "iOS",
		"source": "prebid-mobile"
	},**
	"tid": "6EA27DFD-F221-4D8A-BD38-88C526AF3673",
	"is_debug": true,
	"timeout_millis": 2000,
	"prebid_version": "0.21.0-pre"
}

File: adapters/rubicon/rubicon.go [hunk 2]
```diff
@@ -295,6 +320,11 @@ func (a *RubiconAdapter) Call(ctx context.Context, req *pbs.PBSRequest, bidder *
 		// Assign back our copy
 		rubiReq.User = &userCopy
 
+		deviceCopy := *rubiReq.Device
+		deviceExt := rubiconDeviceExt{RP: rubiconDeviceExtRP{PixelRatio: rubiReq.Device.PxRatio}}
+		deviceCopy.Ext, err = json.Marshal(&deviceExt)
+		rubiReq.Device = &deviceCopy
```
Comment (cindyhere33) [inline adapters/rubicon/rubicon.go]: The existing convention follows creating a copy, making changes and assigning it back.

Can this be:
`deviceExt := rubiconDeviceExt{RP: rubiconDeviceExtRP{PixelRatio: rubiReq.Device.PxRatio}}
rubiReq.Device.Ext, err = json.Marshal(&deviceExt)`

so we can avoid creation of local variable _deviceCopy_. 
Similar comments for _siteCopy_, _rubiconUser_ and _appCopy_ in lines 339, 345, 354.

Comment (sandraleon) [reply adapters/rubicon/rubicon.go]: @cindyhere33 creating a copy avoids race conditions since it points to ref & shared with other adapters

Comment (cindyhere33) [reply adapters/rubicon/rubicon.go]: rubiReq is a local variable. Correct me if I'm wrong.

**Pull request merged**